beacon-chain/blockchain/service.go

```diff
@@ -218,17 +179,126 @@ func (c *ChainService) doesPoWBlockExist(block *types.Block) bool {
+			}
+
+			// Naive fork choice rule: we pick the first block we processed for the previous slot
+			// as canonical.
+			// TODO: Use a better lookup key or abstract into a GetBlock method from DB.
+			hasBlock, err := c.beaconDB.Has(c.blocksPendingProcessing[0])
+			if err != nil {
+				continue
+			}
+			if !hasBlock {
+				continue
+			}
+			encoded, err := c.beaconDB.Get(c.blocksPendingProcessing[0])
+			if err != nil {
+				continue
+			}
+			blockProto := &pb.BeaconBlock{}
+			if err := proto.Unmarshal(encoded, blockProto); err != nil {
+				continue
+			}
+			block := types.NewBlock(blockProto)
+			h, err := block.Hash()
+			if err != nil {
+				log.Errorf("Could not hash incoming block: %v", err)
+			}
+
+			log.Info("Applying fork choice rule")
+			aState := c.chain.ActiveState()
+			cState := c.chain.CrystallizedState()
+			isTransition := cState.IsCycleTransition(c.currentSlot - 1)
+
+			if isTransition {
+				cState, err = cState.CalculateNewCrystallizedState(aState, block.SlotNumber())
+				if err != nil {
+					log.Errorf("Initialize new cycle transition failed: %v", err)
+				}
+			}
+
+			parentBlock, err := c.chain.getBlock(block.ParentHash())
+			if err != nil {
+				log.Errorf("Failed to get parent slot of block 0x%x", h)
+				continue
+			}
+			aState, err = aState.CalculateNewActiveState(block, cState, parentBlock.SlotNumber())
+			if err != nil {
+				log.Errorf("Compute active state failed: %v", err)
+				continue
+			}
+
+			if err := c.chain.SetActiveState(aState); err != nil {
+				log.Errorf("Write active state to disk failed: %v", err)
+			}
+
+			if err := c.chain.SetCrystallizedState(cState); err != nil {
+				log.Errorf("Write crystallized state to disk failed: %v", err)
+			}
+
+			vals, err := casper.ShuffleValidatorsToCommittees(
+				cState.DynastySeed(),
+				cState.Validators(),
+				cState.CurrentDynasty(),
+				cState.CrosslinkingStartShard(),
+			)
+
+			if err != nil {
+				log.Errorf("Unable to get validators by slot and by shard: %v", err)
+				return
+			}
+			log.Debugf("Received %d validators by slot", len(vals))
+
+			// Save canonical block hash with slot number to DB.
+			if err := c.chain.saveCanonicalSlotNumber(block.SlotNumber(), h); err != nil {
+				log.Errorf("Unable to save slot number to db: %v", err)
+			}
+
+			// Save canonical block to DB.
+			if err := c.chain.saveCanonicalBlock(block); err != nil {
+				log.Errorf("Unable to save block to db: %v", err)
+			}
+
+			log.WithField("blockHash", fmt.Sprintf("0x%x", h)).Info("Canonical block determined")
+
+			// We fire events that notify listeners of a new block (or crystallized state in
+			// the case of a state transition). This is useful for the beacon node's gRPC
+			// server to stream these events to beacon clients.
+			if isTransition {
+				c.canonicalCrystallizedStateFeed.Send(cState)
+			}
+			c.canonicalBlockFeed.Send(block)
+
+			// Clear the blocks pending processing.
+			c.blocksPendingProcessing = [][]byte{}
```

Comment:
We need a lock around this line and `updateHead` since they're both accessing the same field in different goroutines.